components/http/okHttp/src/main/java/com/microsoft/kiota/http/OkHttpRequestAdapter.java

```diff
@@ -878,32 +875,24 @@ private void setBaseUrlForRequestInformation(@Nonnull final RequestInformation r
             spanForAttributes.setAttribute(SemanticAttributes.SERVER_ADDRESS, requestURL.getHost());
             spanForAttributes.setAttribute(SemanticAttributes.URL_SCHEME, requestURL.getProtocol());
 
-            RequestBody body =
-                    requestInfo.content == null
-                            ? null
-                            : new RequestBody() {
-                                @Override
-                                public MediaType contentType() {
-                                    final Set<String> contentTypes =
-                                            requestInfo.headers.containsKey(contentTypeHeaderKey)
-                                                    ? requestInfo.headers.get(contentTypeHeaderKey)
-                                                    : new HashSet<>();
-                                    if (contentTypes.isEmpty()) {
-                                        return null;
-                                    } else {
-                                        final String contentType =
-                                                contentTypes.toArray(new String[] {})[0];
-                                        spanForAttributes.setAttribute(
-                                                "http.request_content_type", contentType);
-                                        return MediaType.parse(contentType);
-                                    }
-                                }
-
-                                @Override
-                                public void writeTo(@Nonnull BufferedSink sink) throws IOException {
-                                    sink.writeAll(Okio.source(requestInfo.content));
-                                }
-                            };
+            RequestBody body = null;
```

Comment:
I think we should revert how we build the request body.
HTTP-wise there are two cases for the request body and the content length header:
- the body's length is known in advance, and should be set in the header, so the server knows "when to stop reading". This is typically the case of a JSON payload that comes from serializing models etc...
- the body's length is not known in advance, in which case the header should not be set. The client will be streaming the bytes and using special semantics in the body to indicate it's done streaming the bytes. The is typically the case of a large file upload (not referring to the Microsoft Graph Task), and this is done this way because we CAN'T know in advance the size, and don't want to load everything in memory to avoid memory pressure.

We do have two requirements that come from this:
- setting the telemetry attribute for the content length: We can do this by [reading the content length field of the request body](https://square.github.io/okhttp/3.x/okhttp/okhttp3/RequestBody.html#contentLength--) and if it's not -1, set it, otherwise, ignore it. The RequestBody WriteTo method will ensure the content length is provided when available, exactly like [the dotnet implementation](https://github.com/microsoft/kiota-http-dotnet/blob/30d1d1f70e06498735feba01e7aba71b4271409b/src/HttpClientRequestAdapter.cs#L569)
- providing the input stream to the request body, so it can be streamed, and so it can provide the content length information both in the request it's sending, and to the prior point. This was already implemented with the writeTo method implementation as far as I understand.

Now, it it possible the content length information is never available, but I'd rule it a bad implementation from ok http, and even if that is the case, all of our serialization writers actually return https://docs.oracle.com/javase/8/docs/api/java/io/ByteArrayInputStream.html under the covers, We could downcast the input stream to then read the available method's return, which should give us the right value.

I hope this clarifies things for everyone?

Reply:
additionally, we might also want to read the Content-Type request information header (from the headers collection) when present.

Reply:
@baywet I am having trouble understanding if the contentLength() method is called before or after the writeTo() method, the documentation isn't super clear and it sort of reads as if it is done prior to writeTo(). I think downcasting could work, but that should perhaps be a back up plan. As you mentioned, given our implementation for largeFileUpload we do include the contentLength header, so we should check that value if present, return it if it is, and downcast and check available if that value is not included. Would that make sense?

Reply:
after the writeTo because the value will only be populated once the stream has been enumerated as far as I can tell.
yes I guess the rationale should be:
1. try to read the content-length header from the request information, if present override the implementation of the content length method in request body (if possible)
2. read the content length method only once the request body object has been built, if we get any other value than -1, set it for telemetry.